### PR TITLE
Assignment5- Matthew adams and Rachael DeBoer

### PR DESCRIPTION
**Pull request by MattA41**
@MattA41 and @rkdeboer

### CODE REVIEW

**Comment by SaaWest** [comment]
I will review this

File: GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        //Since there is nothing to reference them, we do not need to worry about what happens to them. They are taken care of in garbage collection
+    }
+    public bool Exists(T value)
+    {
+        ArgumentNullException.ThrowIfNull(value);
+        Node<T> dummy = new(value)
+        {
+            Next = this
+        };
+        while (dummy.Next.Next != this)
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:97]: I believe that this is skipping nodes and not checking all the nodes

File: GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            if (dummy.Next.Value != null)
+            {
+                if (dummy.Next.Value.Equals(value))
+                {
+                    return true;
+                }
+            }
+            dummy.Next = dummy.Next.Next;
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:106]: i don't think this is necessary

File: GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+        }
+        if (dummy.Next.Value != null)
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:109]: Why is this here if the above code is checking the same conditions

File: GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        Node<T> temp = current.Next;
+
+        while (temp != current)
+        {
+            Node<T> nextTemp = temp.Next;
+            temp.Next = this; // Disconnect the node
+            temp = nextTemp;
+        }
+
+        current.Next = current;
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:85]: I think this is all you need to have it clear the rest of the nodes I don't believe the others are necessary.

File: GenericsHomework/Node.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        return ret;
+    }
+    public void Append(T value)
+    {
+        ArgumentNullException.ThrowIfNull(value);
+        if (Exists(value))
+        {
+            return;
+        }
+        Node<T> dummy = new(value)
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:59]: I'm a little confused by lines 59 to 66 why set Next to this instance of the object and why use dummy.Next.Next != this instead of dummy.Next != this

File: GenericsHomework/Node.cs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            Next = this
+        };
+        while (dummy.Next.Next != this)
+        {
+            dummy.Next = dummy.Next.Next;
+        }
+
+        Node<T> newNode = new(value);
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:68]: good job here

File: GenericsHomework/Node.cs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    public T Value { get; }
+    public Node<T> Next { get; private set; }
+
+    public Node(T value)
+    {
+        Value = value;
+        Next = this;
+    }
+
+    public void SetNext(Node<T> next)
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:19]: this can be commented out or delete since its never used

File: GenericsHomework/Node.cs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    {
+
+        Node<T> current = this;
+        Node<T> temp = current.Next;
+
+        while (temp != current)
+        {
+            Node<T> nextTemp = temp.Next;
+            temp.Next = this; // Disconnect the node
+            temp = nextTemp;
```
Comment (SaaWest) [inline GenericsHomework/Node.cs:82]: I think this should be nextTemp.Next = temp that way they circle back on each other

**Review by SaaWest — COMMENTED** [review]
Instructions

    Create a class library project called "GenericsHomework.". ✔
    Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
        Define the Node class
        The values should be homogeneous.
        Include a constructor that takes a value. (No validation is necessary on the value). ✔
        Add a ToString() override that writes out the value's ToString() result. ✔
        Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
            The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
            The Next property setter should be private. ✔
        Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
        Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
            Whether it is sufficient to only set Next to itself ✔
            Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ❌
            Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
        Create an Exists method to test to see if a value exists in the list. ✔
        Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
    You should not rely on any BCL generic classes for your implementation. ✔

Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

    Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌

    Each circle contains n items and each item can belong to one or more Circle instances.
    You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
    You are welcome to use existing BCL generic classes for the extra credit.

    Implement System.Collections.Generic.ICollection<T> on the Node class ❌

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.
    Place all shared project items into a Directory.Build.targets file.
    nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ❌

File: GenericsHomework/Node.cs (hunk 1 above)
Comment (rkdeboer) [reply GenericsHomework/Node.cs:97]: I believe that instead of having a dummy node that just points to this, our dummy points to this in its next variable. This means that when we are checking .Next.Next it will be checking the next node since we start from .Next if that makes sense

File: GenericsHomework/Node.cs (hunk 2 above)
Comment (rkdeboer) [reply GenericsHomework/Node.cs:106]: Care to elaborate? This is what pushes the next node further I dont know what you mean

File: GenericsHomework/Node.cs (hunk 5 above)
Comment (rkdeboer) [reply GenericsHomework/Node.cs:59]: I believe my partner @MattA41  can provide more insight on that but I believe it was because he wanted to ensure that we dont mess too much with the order of the list but have the dummy node using .next so the values are not changing.

File: GenericsHomework/Node.cs (hunk 6 above)
Comment (rkdeboer) [reply GenericsHomework/Node.cs:68]: Thank you!

File: GenericsHomework/Node.cs (hunk 3 above)
Comment (rkdeboer) [reply GenericsHomework/Node.cs:109]: We were having some issues with warnings about Value being null. This may not be relevant code but I can see if it needs changing.

Comment (rkdeboer) [reply GenericsHomework/Node.cs:109]: Just checked it. Not having that check results in an error about the possibility of value being null so its needed to ensure that a null value is not being added

File: GenericsHomework/Node.cs (hunk 7 above)
Comment (BenjaminMichaelis) [reply GenericsHomework/Node.cs:19]: You have access to Next though? why do you need this?

File: GenericsHomework/Node.cs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            dummy.Next = dummy.Next.Next;
+
+        }
+        if (dummy.Next.Value != null)
+        {
+
+            ret += dummy.Next.Value.ToString();
+        }
+
+        return ret;
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:50]:
```suggestion
        if (Value == null) return null;
        return Value.ToString();
```
much simpler way

File: GenericsHomework/Node.cs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    public void Append(T value)
+    {
+        ArgumentNullException.ThrowIfNull(value);
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:54]: Why can't T be null?

File: GenericsHomework/Node.cs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        if (Exists(value))
+        {
+            return;
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:57]: This doesn't allow the caller to know if something has been appended or not.

File: GenericsHomework/Node.cs [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            Next = this
+        };
+        while (dummy.Next.Next != this)
+        {
+            dummy.Next = dummy.Next.Next;
+        }
+
+        Node<T> newNode = new(value);
+        dummy.Next.Next = newNode;
+        newNode.Next = this;
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:70]:
```suggestion
     if (Exists(value)) throw new ArgumentException("This item already exists in the list");
     Node<T> newNode = new(value)
     {
         Next = Next
     };
     Next = newNode;
     return newNode;
```
My suggested implementation

File: GenericsHomework/Node.cs (hunk 4 above)
Comment (BenjaminMichaelis) [reply GenericsHomework/Node.cs:85]: Even just `Next = this;` in that approach

File: GenericsHomework/Node.cs [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+        }
+        if (dummy.Next.Value != null)
+        {
+            if (dummy.Next.Value.Equals(value))
+            {
+                return true;
+            }
+        }
+        return false;
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:116]:
```suggestion
        var node = this;
        do
        {
            if (EqualityComparer<T>.Default.Equals(node.Value, value)) return true;
            node = node.Next;
        } while (node != this);
        return false;
```
This seems over complicated and error prone. And what about if value or Value is null? Why not something like this?
